### ISSUE TITLE
Masschusettts (MA) SID files are all missing ZIP and only have ZIP3

### ISSUE THREAD

**Issue by ran-codes** [Drexel-UHC/hcup-dbt#24]
**MA SID files are missing ZIP and only have ZIP3.**

## Action items

- [ ] figure out if we can link ZIP3 to ZCTA
- [ ] implement in v0.3

### ChatGPT's take

> The HCUP State Inpatient Databases (SID) contain information on patient hospital stays from participating states, including the Massachusetts (MA) SID. The CORE file in the MA SID includes information on patient demographics, diagnoses, procedures, charges, and length of stay, among other variables.
> 
> Regarding ZIP codes, the MA SID CORE file only includes ZIP3 codes, which are the first three digits of the five-digit ZIP code. This is because the MA SID is subject to strict data use agreements and privacy protections under Massachusetts state law, which limits the level of geographic detail that can be reported in order to protect patient confidentiality.
> 
> ZIP3 codes provide a less granular but still useful geographic identifier, as they represent a larger geographic area than the full ZIP code. ZIP3 codes are typically used for mapping and spatial analyses, and can be used to identify the general region or area of the state where the patient resides or was admitted to the hospital.
> 
> In summary, the main difference between ZIP and ZIP3 codes is the level of geographic detail provided. ZIP codes provide a more detailed geographic identifier, while ZIP3 codes represent a larger geographic area. In the case of the MA SID CORE file, only ZIP3 codes are included due to privacy protections under Massachusetts state law.

**Comment by usamabilal** [comment]
Zip3s are huge. Philly has 1 of them: https://www.mapofzipcodes.com/zip-code-map-downloads/pa-pennsylvania-zip-code-map.pdf

From: Ran Li ***@***.***>
Sent: Monday, April 10, 2023 4:26 PM
To: Drexel-UHC/hcup-dbt ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [Drexel-UHC/hcup-dbt] Masschusettts (MA) SID files are all missing ZIP and only have ZIP3 (Issue #24)


External.

MA SID files are missing ZIP and only have ZIP3.

Action items

  *   [ ] figure out if we can link ZIP3 to ZCTA
  *   [ ] implement in v0.3

ChatGPT's take

The HCUP State Inpatient Databases (SID) contain information on patient hospital stays from participating states, including the Massachusetts (MA) SID. The CORE file in the MA SID includes information on patient demographics, diagnoses, procedures, charges, and length of stay, among other variables.

Regarding ZIP codes, the MA SID CORE file only includes ZIP3 codes, which are the first three digits of the five-digit ZIP code. This is because the MA SID is subject to strict data use agreements and privacy protections under Massachusetts state law, which limits the level of geographic detail that can be reported in order to protect patient confidentiality.

ZIP3 codes provide a less granular but still useful geographic identifier, as they represent a larger geographic area than the full ZIP code. ZIP3 codes are typically used for mapping and spatial analyses, and can be used to identify the general region or area of the state where the patient resides or was admitted to the hospital.

In summary, the main difference between ZIP and ZIP3 codes is the level of geographic detail provided. ZIP codes provide a more detailed geographic identifier, while ZIP3 codes represent a larger geographic area. In the case of the MA SID CORE file, only ZIP3 codes are included due to privacy protections under Massachusetts state law.

-

**Comment by her43** [comment]
Idk that I would call this a bug 🤷 . The documentation online shows it's not available. Here are summary tables I made a while ago.

[Variables by state and year.xlsx](https://github.com/Drexel-UHC/hcup-dbt/files/11193799/Variables.by.state.and.year.xlsx)
 See sheet named "ZIP"

**Comment by her43** [comment]
>  figure out if we can link ZIP3 to ZCTA

## IDEA

So, in the majority of cases, the 5-digit-coded ZCTA consist of 1 populated ZIP code with the same 5-digit code, and maybe a handful of PO boxes or "unique" locations, like large buildings. So we wouldn't be able to map directly to a single ZCTA if we only had the first 3 digits.

However, we might be able to come up with a new solution. Here's an idea:

Apparently ([link](https://www.businessinsider.com/what-do-zip-codes-mean-2015-6)), the distinct digits of the 5-digit ZIP code have a meaning, and you can use the first 3 digits to pinpoint a general region.

We could create larger aggregate regions of all ZIP Codes that have the first 3 digits in common.

I checked the ZIP-ZCTA crosswalk (2009-2021) for Massachusetts, and it seems like in general, the first 3 digits of the ZIP code match the first 3 digits of the ZCTA with a few exceptions.

```
# CHECKING FIRST 3 DIGITS OF ZIP CODES VS ZCTA IN CROSSWALK IN MASSACHUSETTS
library(HERcrosswalks)
ma_zip_zcta <- zip_zcta_cw_fun()
ma_zip_zcta_3dig <- ma_zip_zcta %>%
  dplyr::filter(state %in% "MA") %>%
  dplyr::mutate(zip3 = substr(zip, 1, 3),
                zcta3 = substr(zcta, 1, 3),
                match_3 = zip3 == zcta3) 
check_match <- ma_zip_zcta_3dig %>%
  dplyr::group_by(zip3, zcta3, match_3) %>%
  dplyr::summarize(n_zip = length(unique(zip)),
                   n_zcta = length(unique(zcta)) ) %>%
  ungroup() %>% as.data.frame()
print(check_match)

#    zip3 zcta3 match_3 n_zip n_zcta
# 1   010   010    TRUE    70     63
# 2   011   011    TRUE    22     13
# 3   012   012    TRUE    37     32
# 4   013   013    TRUE    31     29
# 5   014   014    TRUE    27     22
# 6   015   015    TRUE    54     45
# 7   016   016    TRUE    18     11
# 8   017   017    TRUE    31     27
# 9   018   018    TRUE    46     32
# 10  018   019   FALSE     1      1
# 11  019   019    TRUE    38     31
# 12  020   020    TRUE    38     26
# 13  021   021    TRUE    62     52
# 14  022   021   FALSE    18      9
# 15  022   022    TRUE     4      3
# 16  023   023    TRUE    45     26
# 17  024   024    TRUE    36     27
# 18  025   025    TRUE    30     25
# 19  026   026    TRUE    40     37
# 20  027   023   FALSE     1      1
# 21  027   027    TRUE    43     36
# 22  055   018   FALSE     2      1
```

**Comment by her43** [comment]
@alinasmahl1

**Comment by her43** [comment]
let's do a crosswalk of zcta (3-digit) to county, see how they compare for MA. how big are these areas?

**Comment by alinasmahl1** [comment]
@her43 just repinging this to see if we addressed this issue.

**Comment by her43** [comment]
We never discussed this any further, but I have some good news:

There is a very clean match up between the first 3 digits of the ZCTA and the county in Massachusetts.
(I have a full R code file looking at this but I can't directly share it here)
```
table_county_zcta3_ma <- 
  table(county = ma_zcta3_county$GEOID,
        zcta3 = ma_zcta3_county$zcta3, useNA = "ifany")
table_county_zcta3_ma
#        zcta3
# county  010 011 012 013 014 015 016 017 018 019 020 021 022 023 024 025 026 027
#   25001   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   1   0
#   25003   1   0   1   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0
#   25005   0   0   0   0   0   0   0   0   0   0   1   0   0   1   0   0   0   1
#   25007   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   1
#   25009   0   0   0   0   0   0   0   0   1   1   0   0   0   0   0   0   0   0
#   25011   1   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0
#   25013   1   1   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0
#   25015   1   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
#   25017   0   0   0   0   1   0   0   1   1   1   0   1   0   0   1   0   0   0
#   25019   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0
#   25021   0   0   0   0   0   0   0   0   0   0   1   1   1   1   1   0   0   1
#   25023   0   0   0   0   0   0   0   0   0   0   1   0   0   1   0   1   0   1
#   25025   0   0   0   0   0   0   0   0   0   0   0   1   1   0   1   0   0   0
#   25027   1   0   0   1   1   1   1   1   0   0   0   0   0   0   0   0   0   0
table(table_county_zcta3_ma)
# table_county_zcta3_ma
#   0   1 
# 207  45 
```

**Comment by her43** [comment]
compare zcta3-digit to NECTA
make sure that patient county is in the data sets (MA) - worst case, we use county for analysis

**Comment by her43** [comment]
Comparing NECTA and the first 3 digits of ZCTA:
What this table shows is 
 - each row is a ZCTA
 - each column is the first 3 digits of the ZCTA
 - each cell value is the proportion of the 3-digit ZCTA area that overlaps with each NECTA

we can see that some ZCTA3 areas are only matched to 1 NECTA, but we can also see several that overlap with multiple NECTAs.
```

> print(ma_zcta3_necta_ptable)  
   necta                                               nectaname   010 011   012   013   014   015 016 017 018 019   020 021 022 023 024   025  026   027 028 999
1  70300                   Amherst Center, MA Micropolitan NECTA 0.091  NA    NA 0.044    NA    NA  NA  NA  NA  NA    NA  NA  NA  NA  NA    NA   NA    NA  NA  NA
2  78100                   Springfield, MA-CT Metropolitan NECTA 0.876   1 0.003 0.094    NA    NA  NA  NA  NA  NA    NA  NA  NA  NA  NA    NA   NA    NA  NA  NA
3  79600                     Worcester, MA-CT Metropolitan NECTA 0.016  NA    NA    NA 0.021 0.937   1  NA  NA  NA    NA  NA  NA  NA  NA    NA   NA    NA  NA  NA
4  99999                                                   99999 0.017  NA 0.199 0.077    NA 0.002  NA  NA  NA  NA    NA  NA  NA  NA  NA 0.228 0.02    NA  NA  NA
5  76150                   North Adams, MA-VT Micropolitan NECTA    NA  NA 0.247 0.002    NA    NA  NA  NA  NA  NA    NA  NA  NA  NA  NA    NA   NA    NA  NA  NA
6  76600                       Pittsfield, MA Metropolitan NECTA    NA  NA 0.551    NA    NA    NA  NA  NA  NA  NA    NA  NA  NA  NA  NA    NA   NA    NA  NA  NA
7  70450                            Athol, MA Micropolitan NECTA    NA  NA    NA 0.265    NA    NA  NA  NA  NA  NA    NA  NA  NA  NA  NA    NA   NA    NA  NA  NA
8  73300                       Greenfield, MA Micropolitan NECTA    NA  NA    NA 0.496    NA    NA  NA  NA  NA  NA    NA  NA  NA  NA  NA    NA   NA    NA  NA  NA
9  74500     Leominster-Fitchburg-Gardner, MA Metropolitan NECTA    NA  NA    NA 0.020 0.690    NA  NA  NA  NA  NA    NA  NA  NA  NA  NA    NA   NA    NA  NA  NA
10 71650       Boston-Cambridge-Quincy, MA-NH Metropolitan NECTA    NA  NA    NA    NA 0.289 0.028  NA   1   1   1 0.959   1   1   1   1    NA   NA 0.193  NA   1
11 77200 Providence-Fall River-Warwick, RI-MA Metropolitan NECTA    NA  NA    NA    NA    NA 0.033  NA  NA  NA  NA 0.041  NA  NA  NA  NA    NA   NA 0.464   1  NA
12 70900                  Barnstable Town, MA Metropolitan NECTA    NA  NA    NA    NA    NA    NA  NA  NA  NA  NA    NA  NA  NA  NA  NA 0.770 0.98 0.009  NA  NA
13 75550                      New Bedford, MA Metropolitan NECTA    NA  NA    NA    NA    NA    NA  NA  NA  NA  NA    NA  NA  NA  NA  NA 0.001   NA 0.333  NA  NA
```


Code that created this table
```
# CHECKING ZCTA AGAINST NECTA AND NECTA DIVISIONS

raw_ma_zcta_necta <- 
  read.csv(file = "E:/HER extra work files/geocorr2014_ma_zcta5_necta.csv")
raw_ma_zcta_necta <- raw_ma_zcta_necta[-1,]
# IT LOOKS LIKE MOST ZCTA MATCH TO 1 NECTA, BUT
# EVEN A VISUAL CHECK SHOWS THIS IS NOT A GOOD MATCH FOR ZIP3/ZCTA3
zcta3_necta_base <- raw_ma_zcta_necta %>%
  mutate(zcta3 = substr(zcta5, 1, 3)) %>%
  group_by(zcta3, necta, nectaname) %>%
  dplyr::summarize(overlap_pop = sum(as.numeric(pop10), na.rm = T),
                   necta_prop_overlap = sum(as.numeric(AFACT2), na.rm = T)) %>%
  ungroup()
zcta3_pops <- raw_ma_zcta_necta %>%
  mutate(zcta3 = substr(zcta5, 1, 3)) %>%
  group_by(zcta3) %>%
  dplyr::summarize(zcta3_total_pop = sum(as.numeric(pop10), na.rm = T))
ma_zcta3_necta <- merge(zcta3_pops,
                        zcta3_necta_base,
                        by = 'zcta3')
ma_zcta3_necta$zcta3_prop_overlap <- 
  round(ma_zcta3_necta$overlap_pop/ma_zcta3_necta$zcta3_total_pop, 3)
ma_zcta3_necta_ptable <- ma_zcta3_necta %>%
  tidyr::pivot_wider(id_cols = c(necta, nectaname),
                     names_from = zcta3,
                     values_from = zcta3_prop_overlap) %>%
  as.data.frame()
print(ma_zcta3_necta_ptable)  
```

**Comment by her43** [comment]
I also checked whether patient county is available in the MA HCUP data - yes it is. The variable is [PSTCO2](https://hcup-us.ahrq.gov/db/vars/siddistnote.jsp?var=pstco2)

**Comment by her43** [comment]
@alinasmahl1 - comment about not getting MA

**Comment by alinasmahl1** [comment]
We're pausing on purchasing MA because we don't have zip code level so can't do any neighborhood level analysis + we it's quite expensive. We may purchase at a later point.
